partiql-eval/src/eval.rs

```diff
@@ -0,0 +1,469 @@
+    ) -> Self {
+        EvalUnpivot {
+            expr,
+            as_key: as_key.to_string(),
+            at_key: at_key.to_string(),
+            output,
+        }
+    }
+}
+
+impl TupleSink for EvalUnpivot {
+    fn push_tuple(&mut self, bindings: Tuple, ctx: &dyn EvalContext) {
+        let result = self.expr.evaluate(&bindings, ctx);
+
+        let tuple = match result {
+            Value::Tuple(tuple) => *tuple,
+            other => other.coerce_to_tuple(),
+        };
+
+        let unpivoted = tuple.0.into_iter().map(|(k, v)| {
+            Tuple::from([(self.as_key.as_str(), v), (self.at_key.as_str(), k.into())])
+        });
+        for t in unpivoted {
+            self.output.push_tuple(t, ctx);
+        }
+    }
+}
+
+impl Evaluable for EvalUnpivot {
+    fn evaluate(&mut self, ctx: &dyn EvalContext) {
+        let empty = Tuple(HashMap::new());
+        self.push_tuple(empty, ctx);
+    }
+}
+
+#[derive(Debug)]
+pub struct EvalWhere {
+    pub expr: Box<dyn EvalExpr>,
+    pub output: Box<dyn TupleSink>,
+}
+
+impl EvalWhere {
+    pub fn new(expr: Box<dyn EvalExpr>, output: Box<dyn TupleSink>) -> Self {
+        EvalWhere { expr, output }
+    }
+
+    #[inline]
+    fn eval_filter(&self, bindings: &Tuple, ctx: &dyn EvalContext) -> bool {
+        let result = self.expr.evaluate(bindings, ctx);
+        match result {
+            Boolean(bool_val) => bool_val,
+            _ => panic!("invalid filter -- not boolean"),
+        }
+    }
+}
+
+impl TupleSink for EvalWhere {
+    fn push_tuple(&mut self, bindings: Tuple, ctx: &dyn EvalContext) {
+        if self.eval_filter(&bindings, ctx) {
+            self.output.push_tuple(bindings, ctx);
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EvalSelect {
+    pub exprs: HashMap<String, Box<dyn EvalExpr>>,
+    pub output: Box<dyn TupleSink>,
+}
+
+impl EvalSelect {
+    pub fn new(exprs: HashMap<String, Box<dyn EvalExpr>>, output: Box<dyn TupleSink>) -> Self {
+        EvalSelect { exprs, output }
+    }
+}
+
+impl TupleSink for EvalSelect {
+    fn push_tuple(&mut self, bindings: Tuple, ctx: &dyn EvalContext) {
+        let proj = self
+            .exprs
+            .iter()
+            .map(|(alias, expr)| (alias.to_string(), expr.evaluate(&bindings, ctx)))
+            .collect();
+        let out = Tuple(proj);
+        self.output.push_tuple(out, ctx)
+    }
+}
+
+#[derive(Debug)]
+pub struct EvalDistinct {
+    pub seen: HashSet<Tuple>,
+    pub output: Box<dyn TupleSink>,
+}
+
+impl EvalDistinct {
+    pub fn new(output: Box<dyn TupleSink>) -> Self {
+        let seen = HashSet::new();
+        EvalDistinct { seen, output }
+    }
+}
+
+impl TupleSink for EvalDistinct {
+    fn push_tuple(&mut self, bindings: Tuple, ctx: &dyn EvalContext) {
+        let is_new = self.seen.insert(bindings.clone());
+        if is_new {
+            self.output.push_tuple(bindings, ctx)
+        }
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct EvalOutputAccumulator {
```

Comment:
Might be a naive question but what is `EvalOutputAccumulator` for? Also, why does is it's only field a `Bag`? Couldn't there be other output types emitted by PartiQL?

Reply:
The PoC only models Bindings -> Bindings relations properly.

We will need to build out Bindings -> Value (i.e., Select Value) and more properly model Value -> Bindings (i.e., From, Unpivot) .

Yes, a Value -> Value expression (e.g. `1+2`) at the head will return a single value, not a bag, but is not modelled currently.